src/pybind/mgr/orchestrator.py

```diff
@@ -0,0 +1,266 @@
+
+"""
+ceph-mgr orchestrator interface
+
+This is a DRAFT for discussion.
+
+Goal: enable UI workflows for cluster service management
+      (such as creating OSDs, in addition to stateless services)
+      using common concepts that are implemented by
+      diverse backends such as Rook, DeepSea, ceph-ansible
+
+Concepts:
+    "Stateful service": a daemon that uses local storage, such as OSD or mon.
+    "Stateless service": a daemon that doesn't use any local storage, such
+                         as an MDS, RGW, nfs-ganesha, iSCSI gateway.
+    "Label": arbitrary string tags that may be applied by administrators
+             to nodes.  Typically administrators use labels to indicate
+             which nodes
+
+Constraints:
+    1. The orchestrator is to be the source of truth for
+       all the physical information, and will be queried directly
+       as needed (i.e. no in-Ceph database of hardware etc).
+    2. The orchestrator handles placement of collections of stateless
+       services.
+    3. The orchestrator accepts explicit placement of individual stateful
+       services, and optionally also accepts label-based automatic placement.
+       (i.e. it *must* support "create OSD at host1:/dev/sdb", and it *may*
+        support "create OSDs on hosts with label=ceph-osd")
+    4. Ceph is not responsible for installing anything on nodes, or hooking
+       up hosts to the orchestrator: nodes and drives only exist to Ceph
+       once they're exposed by the orchestrator.
+
+Flexible features:
+    1. Orchestrator may be aware of one or more "special features" for
+       block devices, such as block-level encryption.  These special
+       features may be reported during discovery or enabled during
+       OSD creation.
+    2. Orchestrator may not have a concept of labels: if it doesn't,
+       any label-filtered calls will match everything.
+
+Excluded functionality:
+    1. No support for multipathed drives: all block devices are to be
+    reported from one host only.
+    2. No networking inventory or configuration.
+
+This is a DRAFT for discussion.
+"""
+
+
+class Orchestrator(object):
+    """
+    All calls in this class are blocking on network IO and may
+    take noticeable length of time to run.  Wrap calls appropriately
+    in long-running task infrastructure when calling from interactive
```

Comment:
At least Salt, and maybe other management tools, already provide asynchronous operation, shouldn't we provide asynchronous versions of each of the `Orchestrator` methods? Maybe expose a public method that can tell us if the `Orchestrator` implementation supports asynchronous operations, and the decide which one to use?

For the asynchronous version of a method I would choose the model of passing a callback function.

Reply:
I'd prefer to keep the public interface to a single style, but I don't necessarily mind if that style is an async one, where any blocking implementation just gets wrapped in threads.

Reply:
I had an informative conversation with @swiftgist. I think maybe @leseb would have some good thoughts here. In DeepSea, and I imagine in Ansible as well, it's more ideal to batch operations together and apply them at one time rather than doing them one-at-a-time to reduce the amount of time spent waiting on the orchestration layer to do its overhead operations. Hypothetically, instead of spending 2 minutes for each call to add an RGW service to 3 nodes (total 6 minutes), if we batched operations, we could spend maybe 3 minutes total. For managing OSDs across the cluster (let's say that 6 OSDs have failed over the weekend and have been physically replaced), 6*2 minutes vs 3 minutes is saving a lot of time.

Maybe we should implement something like an operation queue where add/update/delete commands get collected to be sent to the orchestrator layer all at once.

Reply:
In the RGW example, I would expect that to be a single call to set up a cluster of services (and it's the orchestrator's job to go do things in parallel) rather than a series of calls.

The OSD case is more interesting because we are much more likely to be doing very many individual add_stateful_service calls if the user has done explicit device selection.  If the code calling this interface does many overlapping calls (using an async-style calls), then the implementation does have the option of doing its own batching (something like "call ansible 1 walltime second after the last update call was started").

What about this: all the update operations are asynchronous, they return a completion handle object, and then we have a `Orchestrator.wait([handle, handle...])` function.  In k8s, that wait function would just be waiting on a bunch of API calls to complete, but in Ansible the wait() call would really be the thing that triggers the orchestrator module to take a bunch of enqueued changes and apply them.  I think that probably does the job without needing a full-blown transaction concept?

Reply:
I like @jcsp suggestion of having `Orchestrator.wait()` function. By using an asynchronous interface, the orchestrator implementation can use any batching strategy.

Reply:
Re @rjfd, I think we should assume that all methods are asynchronous. And it would be helpful to request the orchestrator's status (e.g., "Applying changes: <stuff>. 65% done", or whatever). We may also need to handle the case where the orchestrator is busy and cannot take new input. Status might then be "Applying changes: <stuff>. 65% done. Ready_for_input? No".